Improve error message when uploading image without correct extension or mimetype

jordisala1991 opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

<!-- Describe your Pull Request content here -->

<!--
    Show us you choose the right branch.
    Different branches are used for different things :
    - 4.x is for everything backwards compatible, like patches, features and deprecation notices
    - 5.x is for deprecation removals and other changes that cannot be done without a BC-break
    More details here: https://github.com/sonata-project/SonataMediaBundle/blob/4.x/CONTRIBUTING.md#base-branch
-->
I am targeting this branch, because this is a bugfix.

<!--
    Specify which issues will be fixed/closed.
    Remove it if this is not related.
-->

Closes #2208 .

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataMediaBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Fixed
- Fixed call to `strlen()` with possible null value.
- Improved error message for uploaded images that do not have an allowed extension or mime type.
```

jordisala1991 committed:
Add test for update metadata on images

jordisala1991 commented:
@sarim let's see what github actions says, but on my computer I couldn't make this test fail.

sarim commented:
@jordisala1991 I think your test are passing because resizer / provider etc.. service are mocked in that test file, not real services. 

Here I just grabbed a [random cat pic](https://i.imgur.com/kDZqKq2.jpeg) from imgur and can reproduce the error.

![image](https://user-images.githubusercontent.com/1235888/141441759-5ebc302d-d114-4ad5-83e7-7ee0e0db5def.png)

jordisala1991 commented:
You are right @sarim, but I have just changed the test to not mock anything related to size calculations. Now it is not mocked. I have left the GD implementation of imagine because that's what we are installing on github-actions. But I have tested with: Gd, imagick and gmagick. Those 3 implementations works and produce the same result.

We can merge this test because at the end it improves coverage, but as far as I can tell, if there is any issue you need to look at Imagine library, not here.

I left this as patch because I have added a fix on a strlen call, but it is only related to stop mocking things we don't own rather than the original bug reported.

jordisala1991 committed:
Fix call to strlen with null value

jordisala1991 commented:
Now that I take a look at the add media command, do you have the problem only when using the command? did you tried to use the admin panel to upload that image?

I think the problem might be because of the usage of this command.

jordisala1991 commented:
Finally, it is related directly to the command and how all the flow executes. Now github actions fails with the same error reported.

sarim commented:
You kinda lost me in this complex testing technology. If you would've just run the command from any repo (for example I showed screen from sonata sandbox) you would've encountered the problem. The problem is simple, as I described in my original issue. The problem was NOT in sonata v2. I upgraded a symfony2+sonata2 project to symfony 5 + sonata 4 and encountered this problem.

Let me reiterate : From [Usage doc](https://docs.sonata-project.org/projects/SonataMediaBundle/en/3.x/reference/usage/)
```php
$media = new Media();
$media->setBinaryContent(PATH TO IMAGE FILE);
$media->setContext('default');
$media->setProviderName('sonata.media.provider.image');

/** @var Sonata\MediaBundle\Model\MediaManagerInterface $mediaManager */
$mediaManager->save($media);
```

This is the way to create a new media. I was creating image using this code after a form submit, as you know, php saves uploaded files in `/tmp/ABCD_RANDOM_CHARACTERS_WITHOUT_EXTENSION`. So thats the path I'm passing to `$media->setBinaryContent`. Which used to work fine in sonata v2, but now it does not.

Now the error is simple and easy to reproduce, as the media add command uses the same code to create media, same error happens there.

I brought up the add command because it is very to easy to check. You just cd to any sonata project and run the command. The main issue is you can't add media using the referenced code.

jordisala1991 commented:
I need more time to investigate. So far my findings are that it is harder to reproduce on a unit test, but I can clearly reproduce on the functional test, running the command directly. 

Seems related to a missing step during the upload process because I can updateMetadata on a media and it works fine, with or without file extension.

The process to fix this requires a test to ensure no regressions after, so thats why is not as simple as opening a project and running the command you provided.

sarim commented:
First we want to verify if a bug really exist? or a user error. To answer that yes/no question you can run the command to verify, that was my intention. From my POV it looked like you were trying to reproduce it using complex testing technology while running a simple command would do. I mean the issue is still marked `unconfirmed`. 


> The process to fix this requires a test to ensure no regressions after, so thats why is not as simple as opening a project and running the command you provided

Ah, I understand.

jordisala1991 commented:
Until right now I couldn't confirm it exists thats why it does have the label.

jordisala1991 committed:
Throw exception when allowed mimetypes or allowed extensions do not match with the uploaded image

jordisala1991 commented:
The focus of the PR has changed a bit after finding why it doesn't work without extensions, more here: https://github.com/sonata-project/SonataMediaBundle/issues/2208

phansys commented on src/Provider/ImageProvider.php:
```diff
@@ -215,9 +216,28 @@ protected function doTransform(MediaInterface $media)
+            throw new UploadException(sprintf(
+                'The image extension "%s" is not one of the allowed (%s)',
+                $extension,
+                implode(', ', $this->allowedExtensions)
```
```suggestion
                '"'.implode('", "', $this->allowedExtensions).'"'
```

jordisala1991 replied:
This would produce "" for an empty array, if we want to add quotes it needs more thinking

phansys replied:
If this case is possible, I think we could add a previous condition (`[] === $this->allowedExtensions`) with a specific message stating that there are no allowed extensions.

jordisala1991 replied:
Fixed.

phansys commented:
Thank you @jordisala1991!
